pinot-segment-local/src/main/java/org/apache/pinot/segment/local/recordtransformer/ComplexTypeTransformer.java

```diff
@@ -0,0 +1,184 @@
+/**
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.pinot.segment.local.recordtransformer;
+
+import com.google.common.annotations.VisibleForTesting;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import javax.annotation.Nullable;
+import org.apache.pinot.spi.config.table.TableConfig;
+import org.apache.pinot.spi.config.table.ingestion.ComplexTypeHandlingConfig;
+import org.apache.pinot.spi.data.readers.GenericRow;
+
+
+public class ComplexTypeTransformer implements RecordTransformer {
+  private static final CharSequence DELIMITER = ".";
+  private final List<String> _collectionsToUnnest;
+
+  public ComplexTypeTransformer(TableConfig tableConfig) {
+    if (tableConfig.getIngestionConfig() != null
+        && tableConfig.getIngestionConfig().getComplexTypeHandlingConfig() != null) {
+      _collectionsToUnnest =
+          tableConfig.getIngestionConfig().getComplexTypeHandlingConfig().getUnnestConfig() != null ? tableConfig
+              .getIngestionConfig().getComplexTypeHandlingConfig().getUnnestConfig() : new ArrayList<>();
+    } else {
+      _collectionsToUnnest = new ArrayList<>();
+    }
+  }
+
+  @VisibleForTesting
+  public ComplexTypeTransformer(List<String> unnestCollections) {
+    _collectionsToUnnest = new ArrayList<>(unnestCollections);
+    Collections.sort(_collectionsToUnnest);
+  }
+
+  public static boolean isComplexTypeHandlingEnabled(TableConfig tableConfig) {
+    if (tableConfig.getIngestionConfig() == null
+        || tableConfig.getIngestionConfig().getComplexTypeHandlingConfig() == null
+        || tableConfig.getIngestionConfig().getComplexTypeHandlingConfig().getMode() == null) {
+      return false;
+    }
+    return tableConfig.getIngestionConfig().getComplexTypeHandlingConfig().getMode()
+        != ComplexTypeHandlingConfig.Mode.NONE;
+  }
+
+  @Nullable
+  @Override
+  public GenericRow transform(GenericRow record) {
+    flattenMap(record, new HashSet<>(record.getFieldToValueMap().keySet()));
```

Comment:
Why does the `Set<String>` returned by `keySet()` need to be copied into a new `HashSet`?

Reply:
otherwise it'll run into concurrent modification exceptions, as the recursion may update the record

Reply:
I think using an Iterator (instead of for loop) might avoid ConcurrentModificationException. I am mainly concerned about all the copying being done and trying to see if that can be avoided.

Reply:
Can you give an example? ConcurrentModificationException occurs when a collection being iterated over is modified.

Reply:
I see. Thanks for sharing the pointer. I think the rewrite is possible, but at the tradeoff code readability. Particularly, the current logic is already a bit complex, given its use of recursion. I think we can look into it if we identify it as perf bottleneck.

Reply:
Iterative [level-order tree traversal](https://java2blog.com/binary-tree-level-order-traversal-java/) might be another option as well.

Reply:
I'm not sure how much we'll save with the level-order traversal. As it'll populate the items to another linkedlist which effectively simulate `array.close`. IMHO, the shallow copy shall be cheap enough, and the code readability is more important.

Reply:
Maybe other reviewers can chime in here. If not, we can assume that this is not a major concern. Although, at least in my mind, it's the copy that is causing confusion. Because it makes it seem like there is a lot more things going on here other than regular recursion and iteration. At the very least through, a comment saying why the copy is being done would be useful for clarifying things.

Reply:
Since we are going to remove entry and add new entries at the same time, there are 2 ways: 1. use iterator and only remove entries, keep the new entries in a separate map and merge 2 maps in the end; 2. put the keys in a separate list (faster than set for iteration).
Since both approaches require extra data structure, I feel it is okay to just cache the keys (suggest changing to list though).